### PR TITLE
[Async ALP] Make getLeadershipToken return ListenableFuture and chain it to avoid throwing

### PR DESCRIPTION
**Pull request by jeremyk-91**
Note: This PR is pointed at #4610, not at develop.

**Goals (and why)**:
- Normally if I call a method `blahAsync()` returning `ListenableFuture<T>` and the underlying call fails, I expect a failed future, not my async thing throwing an exception.

**Implementation Description (bullets)**:
- Change `getLeadershipToken()` to use `ListenableFuture` and chain it accordingly.

**Testing (What was existing testing like?  What have you done to improve it?)**:
- Added a test that would have thrown the `NCLE` directly in the implementation James and I paired on.
- Test failures happen in the base branch for the paxos consensus fast test.

**Concerns (what feedback would you like?)**:
- At points I just get the value of the future, but I think this makes sense, as when these occur on the async path they're part of future-transforms.

**Where should we start reviewing?**: AwaitingLeadershipProxy?

**Priority (whenever / two weeks / yesterday)**: this week

### CODE REVIEW

File: leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java [hunk 1]
```diff
@@ -225,18 +224,19 @@ protected Object handleInvocation(Object proxy, Method method, Object[] args) th
             return null;
         }
 
-        final LeadershipToken leadershipToken = getLeadershipToken();
+        ListenableFuture<LeadershipToken> leadershipToken = getLeadershipToken();
```
Comment (j-baker) [inline leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java]: I don't know you'd want this - it seems reasonable to throw if you're knowably not the leader

Comment (j-baker) [reply leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java]: but does add some complexity?

Comment (j-baker) [reply leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java]: Idk - it's not something i'd like to be part of the contract, if that makes sense

Comment (jeremyk-91) [reply leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java]: Yep, the main reason for this was that I didn't want to throw in the async case. I did consider the alternative of splitting here to decide whether to throw immediately or return immediateFailedFuture on whether we were on an async path - didn't really like either solution. No strong opinions, wouldn't mind changing if you think that works better.

It does seem that we should handle the exceptions here similar to how we get them in the `isStillLeading` part (that could just be that I know my round is not the latest, for instance), which I would imagine is throwing for sync and returning a failed future for async.

**Comment by jeremyk-91** [comment]
Discussed offline: the original API should be fine, it is similar to Guava `AsyncFunction` in that throwing is allowed and has the same semantics as a `Futures.immediateFailedFuture(e)`.

**Pull request closed without merging**